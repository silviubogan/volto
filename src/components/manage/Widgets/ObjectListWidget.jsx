import { map } from 'lodash';
import {
  Button,
  Form,
  Grid,
  Header,
  Icon,
  Input,
  Label,
  Modal,
  Segment,
} from 'semantic-ui-react';
import React, { Fragment, useState } from 'react';
import PropTypes from 'prop-types';
import { Icon as VoltoIcon } from '@plone/volto/components';
import deleteSVG from '@plone/volto/icons/delete.svg';
import penSVG from '@plone/volto/icons/pen.svg';
import addSVG from '@plone/volto/icons/add.svg';

import ObjectWidget from './ObjectWidget';

export const FlatObjectList = ({ id, value = [], schema, onChange }) => {
  return (
    <div className="objectlist-widget-content">
      {!value && <ObjectWidget schema={schema} />}
      {value.map((obj, index) => (
        // TODO: notice that the Fragment key={} might cause problems, need to test
        <Fragment key={index}>
          <Grid>
            <Grid.Column width={11}>
              <Segment>
                <ObjectWidget
                  key={index}
                  schema={schema}
                  value={obj}
                  onChange={(fi, fv) =>
                    onChange(
                      id,
                      value.map((v, i) =>
                        i !== index ? v : { ...v, [fi]: fv },
                      ),
                    )
                  }
                />
              </Segment>
            </Grid.Column>
            <Grid.Column width={1}>
              <Button.Group>
                <Button
                  basic
                  circular
                  size="mini"
                  onClick={() =>
                    onChange(
                      id,
                      value.filter((v, i) => i !== index),
                    )
                  }
                >
                  {/* TODO: instead of px use rem if possible */}
                  <VoltoIcon size="20px" name={deleteSVG} />
                </Button>
              </Button.Group>
            </Grid.Column>
          </Grid>
        </Fragment>
      ))}
    </div>
  );
};

export const ModalObjectListForm = (props) => {
  const {
    open,
    title,
    className,
    onSave,
    onCancel,
    schema,
    id,
    value = [],
  } = props;

  const empty = {};

  const [stateValue, setStateValue] = useState(value);

  return (
    <Modal open={open} className={className}>
      <Header>{title}</Header>
      <Modal.Content scrolling>
        <FlatObjectList
          {...props}
          value={stateValue}
          onChange={(id, v) => setStateValue(v)}
        />
      </Modal.Content>
      <Modal.Actions>
        <Button
          primary
          basic
          circular
          floated="left"
          size="big"
          className="icon"
          onClick={() => setStateValue([...stateValue, empty])}
        >
          {/* TODO: instead of px use rem if possible */}
          <VoltoIcon size="20px" name={addSVG} />
          Add {schema.title}
        </Button>

        <Button
          basic
          circular
          primary
          floated="right"
          icon="arrow right"
          aria-label="Save"
          title="Save"
          size="big"
          onClick={() => onSave(id, stateValue)}
        />
        <Button
          basic
          circular
          secondary
          icon="remove"
          aria-label="Cancel"
          title="Cancel"
          floated="right"
          size="big"
          onClick={onCancel}
        />
      </Modal.Actions>
    </Modal>
  );
};

const ObjectListWidget = (props) => {
  const [open, setOpen] = useState(false);
  const {
    id,
    value = [],
    schema,
    onChange,
    required,
    error,
    fieldSet,
    title,
    description,
    onDelete,
    onEdit,
  } = props;

  return (
    <>
      <ModalObjectListForm
        {...props}
        open={open}
        onSave={(id, value) => {
          setOpen(false);
          onChange(id, value);
        }}
        onCancel={() => setOpen(false)}
      />
      <Form.Field
        inline
        required={required}
        error={(error || []).length > 0}
        className={description ? 'help text' : 'text'}
        id={`${fieldSet || 'field'}-${id}`}
      >
        <Grid>
          <Grid.Row stretched>
            <Grid.Column width="4">
              <div className="wrapper">
                <label htmlFor={`field-${id}`}>
                  {onEdit && (
                    <i
                      aria-hidden="true"
                      className="grey bars icon drag handle"
                    />
                  )}
                  {title}
                </label>
              </div>
            </Grid.Column>
            <Grid.Column width="8">
              {onEdit && (
                <div className="toolbar">
                  <button
                    className="item ui noborder button"
                    onClick={() => onEdit(id, schema)}
                  >
                    <Icon name="write square" size="large" color="blue" />
                  </button>
                  <button
                    aria-label="Delete"
                    className="item ui noborder button"
                    onClick={() => onDelete(id)}
                  >
                    <Icon name="close" size="large" color="red" />
                  </button>
                </div>
              )}

              <Input
                id={`field-${id}`}
                name={id}
                disabled={true}
                icon={penSVG}
                value={`A collection of ${value.length} items`}
              />
              <button onClick={() => setOpen(true)}>
                {/* TODO: instead of px use rem if possible */}
                <VoltoIcon name={penSVG} size="18px" />
              </button>

              {map(error, (message) => (
                <Label key={message} basic color="red" pointing>
                  {message}
                </Label>
              ))}
            </Grid.Column>
          </Grid.Row>
          {description && (
            <Grid.Row stretched>
              <Grid.Column stretched width="12">
                <p className="help">{description}</p>
              </Grid.Column>
            </Grid.Row>
          )}
        </Grid>
      </Form.Field>
    </>
  );
};

/**
 * Property types.
 * @property {Object} propTypes Property types.
 * @static
 */
ObjectListWidget.propTypes = {
  id: PropTypes.any.isRequired,
  schema: PropTypes.object.isRequired,
  error: PropTypes.any.isRequired,
  value: PropTypes.array,
  onChange: PropTypes.func.isRequired,
  required: PropTypes.bool.isRequired,
  fieldSet: PropTypes.any.isRequired,
  title: PropTypes.string.isRequired,
  description: PropTypes.string.isRequired,
  onDelete: PropTypes.func.isRequired,
  onEdit: PropTypes.func.isRequired,
};

/**
 * Default properties.
 * @property {Object} defaultProps Default properties.
 * @static
 */
ObjectListWidget.defaultProps = {
  value: [],
};

export default ObjectListWidget;
