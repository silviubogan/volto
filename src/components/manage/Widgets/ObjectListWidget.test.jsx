import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-intl-redux';
import { render, fireEvent } from '@testing-library/react';
import '@testing-library/jest-dom/extend-expect';
import ObjectListWidget from './ObjectListWidget';

const mockStore = configureStore();

const LinkSchema = {
  title: 'Link',
  fieldsets: [
    {
      id: 'internal',
      title: 'Internal',
      fields: ['internal_link'],
    },
    {
      id: 'external',
      title: 'External',
      fields: ['external_link'],
    },
    {
      id: 'email',
      title: 'Email',
      fields: ['email_address', 'email_subject'],
    },
  ],
  properties: {
    internal_link: {
      widget: 'object_browser',
      title: 'Internal link',
    },
    external_link: {
      title:
        'External URL (can be relative within this site or absolute if it starts with http:// or https://)',
    },
    email_address: {
      title: 'Email address',
    },
    email_subject: {
      title: 'Email subject (optional)',
    },
  },
  required: [],
};

test('renders an object list widget component', () => {
  const store = mockStore({
    search: {},
    intl: {
      locale: 'en',
      messages: {},
    },
  });

  const component = renderer.create(
    <Provider store={store}>
      <ObjectListWidget
        id="my-widget"
        schema={LinkSchema}
        title="My Widget"
        onChange={() => {}}
        error={{}}
        value={[
          { external_link: 'https://dgg.gg' },
          { external_link: 'https://wikipedia.org' },
        ]}
        required={true}
        fieldSet={{}}
        description="My description"
        onDelete={() => {}}
        onEdit={() => {}}
      />
    </Provider>,
  );

  const json = component.toJSON();
  expect(json).toMatchSnapshot();
});

test('renders an object list widget component and changes its value by clicking a button', async () => {
  const store = mockStore({
    search: {},
    intl: {
      locale: 'en',
      messages: {},
    },
  });

  let valueState = [
    { external_link: 'https://ddg.gg' },
    { external_link: 'https://wikipedia.org' },
  ];

  const { getByText, asFragment, rerender, findByTestId } = render(
    <Provider store={store}>
      <>
        <ObjectListWidget
          id={`my-widget`}
          schema={LinkSchema}
          title="My Widget"
          onChange={(id, v) => {
            valueState = v;
            rerender();
          }}
          error={{}}
          value={valueState}
          required={true}
          fieldSet={{}}
          description="My description"
          onDelete={() => {}}
          onEdit={() => {}}
        />
        <button
          onClick={(e) => {
            valueState = [{ external_link: 'https://duckduckgo.com' }];
            rerender();
            e.preventDefault();
          }}
        >
          Click me !
        </button>
      </>
    </Provider>,
  );

  expect(asFragment()).toMatchSnapshot();

  fireEvent.click(getByText('Click me !'));

  fireEvent.click(await findByTestId('big-pen-button'));

  fireEvent.click(getByText('External'));

  expect(asFragment()).toMatchSnapshot();
});
