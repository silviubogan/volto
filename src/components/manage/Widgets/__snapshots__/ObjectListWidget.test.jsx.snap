// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders an object list widget component 1`] = `
<div
  className="inline required field help text"
  id="[object Object]-my-widget"
>
  <div
    className="ui grid"
  >
    <div
      className="stretched row"
    >
      <div
        className="four wide column"
      >
        <div
          className="wrapper"
        >
          <label
            htmlFor="field-my-widget"
          >
            <i
              aria-hidden="true"
              className="grey bars icon drag handle"
            />
            My Widget
          </label>
        </div>
      </div>
      <div
        className="eight wide column"
      >
        <div
          className="toolbar"
        >
          <button
            className="item ui noborder button"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="blue write square large icon"
              onClick={[Function]}
            />
          </button>
          <button
            aria-label="Delete"
            className="item ui noborder button"
            onClick={[Function]}
          >
            <i
              aria-hidden="true"
              className="red close large icon"
              onClick={[Function]}
            />
          </button>
        </div>
        <div
          className="ui disabled icon input"
        >
          <input
            disabled={true}
            id="field-my-widget"
            name="my-widget"
            onChange={[Function]}
            tabIndex={-1}
            type="text"
            value="A collection of 2 items"
          />
          <i
            aria-hidden="true"
            attributes={
              Object {
                "viewBox": "",
                "xmlns": "",
              }
            }
            className="icon"
            onClick={[Function]}
          />
        </div>
        <button
          onClick={[Function]}
        >
          <svg
            className="icon"
            dangerouslySetInnerHTML={
              Object {
                "__html": undefined,
              }
            }
            onClick={null}
            style={
              Object {
                "fill": "currentColor",
                "height": "18px",
                "width": "auto",
              }
            }
            viewBox=""
            xmlns=""
          />
        </button>
      </div>
    </div>
    <div
      className="stretched row"
    >
      <div
        className="stretched twelve wide column"
      >
        <p
          className="help"
        >
          My description
        </p>
      </div>
    </div>
  </div>
</div>
`;
